// jest-validate/build/utils.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.logValidationWarning =
  exports.formatPrettyObject =
  exports.format =
  exports.createDidYouMeanMessage =
  exports.WARNING =
  exports.ValidationError =
  exports.ERROR =
  exports.DEPRECATION =
    void 0;

function _chalk() {
  const data = _interopRequireDefault(require('chalk'));

  _chalk = function () {
    return data;
  };

  return data;
}

function _leven() {
  const data = _interopRequireDefault(require('leven'));

  _leven = function () {
    return data;
  };

  return data;
}

function _prettyFormat() {
  const data = require('pretty-format');

  _prettyFormat = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj};
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
